pkg/provider/ngrok/client.go

```diff
@@ -0,0 +1,329 @@
+}
+
+type SecretsClient interface {
+	Create(context.Context, *ngrok.SecretCreate) (*ngrok.Secret, error)
+	Delete(context.Context, string) error
+	Get(context.Context, string) (*ngrok.Secret, error)
+	List(*ngrok.Paging) ngrok.Iter[*ngrok.Secret]
+	Update(context.Context, *ngrok.SecretUpdate) (*ngrok.Secret, error)
+}
+
+type client struct {
+	vaultClient   VaultClient
+	secretsClient SecretsClient
+	vaultName     string
+	vaultID       string
+	vaultIDMu     sync.RWMutex
+}
+
+func (c *client) PushSecret(ctx context.Context, secret *corev1.Secret, data esv1.PushSecretData) error {
+	if secret == nil {
+		return errCannotPushNilSecret
+	}
+
+	// First, make sure the vault name still matches the ID we have stored. If not, we have to look it up again.
+	err := c.verifyVaultNameStillMatchesID(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to verify vault name still matches ID: %w", err)
+	}
+
+	// Prepare the secret data for pushing
+	var value []byte
+
+	// If key is specified, get the value from the secret data
+	if data.GetSecretKey() != "" {
+		var ok bool
+		value, ok = secret.Data[data.GetSecretKey()]
+		if !ok {
+			return fmt.Errorf("key %s not found in secret", data.GetSecretKey())
+		}
+	} else { // otherwise, marshal the entire secret data as JSON
+		value, err = json.Marshal(secret.Data)
+		if err != nil {
+			return fmt.Errorf("json.Marshal failed with error: %w", err)
+		}
+	}
+
+	// Calculate the checksum of the value to add to metadata
+	valueChecksum := sha256.Sum256(value)
+
+	psmd, err := parseAndDefaultMetadata(data.GetMetadata())
+	if err != nil {
+		return fmt.Errorf("failed to parse push secret metadata: %w", err)
+	}
+
+	psmd.Metadata["_sha256"] = hex.EncodeToString(valueChecksum[:])
+	metadataJSON, err := json.Marshal(psmd.Metadata)
+	if err != nil {
+		return fmt.Errorf("failed to marshal metadata for ngrok api: %w", err)
+	}
+
+	// Check if the secret already exists in the vault
+	existingSecret, err := c.getSecretByVaultIDAndName(ctx, c.getVaultID(), data.GetRemoteKey())
+	if err != nil {
+		if !errors.Is(err, errVaultSecretDoesNotExist) {
+			return fmt.Errorf("failed to get secret: %w", err)
+		}
+
+		// If the secret does not exist, create it
+		_, err = c.secretsClient.Create(ctx, &ngrok.SecretCreate{
+			VaultID:     c.getVaultID(),
+			Name:        data.GetRemoteKey(),
+			Value:       string(value),
+			Metadata:    string(metadataJSON),
+			Description: psmd.Description,
+		})
+		return err
+	}
+
+	// If the secret exists, update it
+	_, err = c.secretsClient.Update(ctx, &ngrok.SecretUpdate{
+		ID:          existingSecret.ID,
+		Value:       ptr.To(string(value)),
+		Metadata:    ptr.To(string(metadataJSON)),
+		Description: ptr.To(psmd.Description),
+	})
+	return err
+}
+
+func (c *client) SecretExists(ctx context.Context, ref esv1.PushSecretRemoteRef) (bool, error) {
+	err := c.verifyVaultNameStillMatchesID(ctx)
+	if errors.Is(err, errVaultDoesNotExist) {
+		return false, nil
+	} else if err != nil {
```

Comment:
nit: remove this from else as you have an early return in the first if

Reply:
Fixed in https://github.com/external-secrets/external-secrets/pull/5160/commits/2f72b1abcee363d4d18a6a8ff746dbcea3718a0c